epoch.go

```diff
@@ -172,7 +176,8 @@ func (e *Epoch) init() error {
 	e.sched = NewScheduler(e.Logger)
 	e.monitor = NewMonitor(e.StartTime, e.Logger)
 	e.cancelWaitForBlockNotarization = func() {}
-	e.finishCtx, e.finishFn = context.WithCancel(context.Background())
+	e.ctx, e.cancelCtx = context.WithCancel(context.Background())
+	e.blockBuilderCtx, e.blockBuilderCancelFunc = context.WithCancel(e.ctx)
```

Comment:
why initialize the builder context here and then overwrite it again? furthermore I think a context should be cancelled before we overwrite it so doesn't this create a leak?